File: src/main/java/net/flyingfishflash/loremlist/api/data/response/LrmListItemResponse.java
package net.flyingfishflash.loremlist.api.data.response;

import java.util.Set;
import java.util.UUID;
import kotlinx.datetime.Instant;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct;
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItem;

/** Representation of a list item in the context of a list */
public record LrmListItemResponse(
    UUID id,
    String name,
    String description,
    int quantity,
    boolean isSuppressed,
    String owner,
    Instant created,
    String creator,
    Instant updated,
    String updater,
    Set<LrmListSuccinct> lists) {

  public static LrmListItemResponse fromLrmListItem(LrmListItem lrmListItem) {
    return new LrmListItemResponse(
        lrmListItem.id(),
        lrmListItem.name(),
        lrmListItem.description(),
        lrmListItem.quantity(),
        lrmListItem.isSuppressed(),
        lrmListItem.owner(),
        lrmListItem.created(),
        lrmListItem.creator(),
        lrmListItem.updated(),
        lrmListItem.updater(),
        lrmListItem.lists());
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/api/data/response/LrmListItemResponse.kt
package net.flyingfishflash.loremlist.api.data.response

import kotlinx.datetime.Instant
import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import net.flyingfishflash.loremlist.core.serialization.UuidSerializer
import net.flyingfishflash.loremlist.domain.lrmlist.LrmListSuccinct
import net.flyingfishflash.loremlist.domain.lrmlistitem.LrmListItem
import java.util.UUID

/**
 * Representation of a list item in the context of a list
 */
@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class LrmListItemResponse(
  @Serializable(with = UuidSerializer::class)
  val id: UUID,
  val name: String,
  val description: String? = null,
  @EncodeDefault val quantity: Int = 0,
  val isSuppressed: Boolean,
  val owner: String,
  val created: Instant,
  val creator: String,
  val updated: Instant,
  val updater: String,
  val lists: Set<LrmListSuccinct>,
) {
  companion object {
    fun fromLrmListItem(lrmListItem: LrmListItem): LrmListItemResponse {
      return LrmListItemResponse(
        id = lrmListItem.id,
        name = lrmListItem.name,
        description = lrmListItem.description,
        quantity = lrmListItem.quantity,
        isSuppressed = lrmListItem.isSuppressed,
        owner = lrmListItem.owner,
        created = lrmListItem.created,
        creator = lrmListItem.creator,
        updated = lrmListItem.updated,
        updater = lrmListItem.updater,
        lists = lrmListItem.lists,
      )
    }
  }
}
